In [44]:
temps = [24.4, 25.9, 28.6, 19.3, 26.0, 22.1]

fahrenheit = [t * 9/5 + 32 for t in temps]

print(fahrenheit)

[75.92, 78.62, 83.48, 66.74000000000001, 78.8, 71.78]


In [45]:
hot = [t for t in temps if t >= 26]

print(hot)

[28.6, 26.0]


In [46]:
cool_rounded = [round(t) for t in temps if t < 24]

print(cool_rounded)

[19, 22]


In [47]:
cool_rounded = [round(t) for t in temps if t < 24]

print(cool_rounded)

[19, 22]


In [48]:
raw = ["24.4", "25.9", "28.6", "19.3"]

values = [float(x) for x in raw]

print(values)
print("Average:", sum(values) / len(values))

[24.4, 25.9, 28.6, 19.3]
Average: 24.55


In [49]:
#PART 2 Lambda

In [50]:
results = [
    ("Aline", 85),
    ("Eric", 78),
    ("Diane", 91),
    ("Samuel", 92),
    ("Grace", 76)
]

ranked = sorted(results, key=lambda x: x[1], reverse=True)

print(ranked[:3])

[('Samuel', 92), ('Diane', 91), ('Aline', 85)]


In [51]:
rain_log = [
    (3, 0.0),
    (7, 12.5),
    (11, 3.2),
    (14, 26.2),
    (18, 8.9)
]

wettest = max(rain_log, key=lambda x: x[1])

print(wettest)

(14, 26.2)


In [52]:
#PART 3 Numpy

In [53]:
import numpy as np
import time

arr = np.arange(1000000)

start = time.time()

result = []
for x in arr:
    result.append(x * 2)

loop_time = time.time() - start

start = time.time()

result2 = arr * 2

numpy_time = time.time() - start

print("Loop time:", loop_time)
print("NumPy time:", numpy_time)
print("NumPy was approximately", loop_time / numpy_time, "times faster.")

Loop time: 0.645073413848877
NumPy time: 0.007819414138793945
NumPy was approximately 82.49638686465225 times faster.


In [56]:
temps = np.array(temps)

print(f"Mean temperature: {temps.mean():.2f} °C")
print(f"Maximum temperature: {temps.max():.2f} °C")
print(f"Minimum temperature: {temps.min():.2f} °C")
print(f"Standard deviation: {temps.std():.2f} °C")

Mean temperature: 24.38 °C
Maximum temperature: 28.60 °C
Minimum temperature: 19.30 °C
Standard deviation: 2.99 °C


In [57]:
position = np.argmax(temps)

print(f"Hottest day: {dates[position]}")
print(f"Temperature: {temps[position]} °C")

Hottest day: 1970-01-01
Temperature: 28.6 °C


In [58]:
#PART4 

hot_mask = temps >= 26

print("Number of hot days:", hot_mask.sum())
print(f"Average hot-day temperature: {temps[hot_mask].mean():.2f} °C")

Number of hot days: 2
Average hot-day temperature: 27.30 °C


In [59]:
rain_mask = rain > 0

print("Number of rainy days:", rain_mask.sum())
print("Total rainfall:", rain.sum(), "mm")
print("Heaviest single-day rainfall:", rain.max(), "mm")

Number of rainy days: 16
Total rainfall: 331.20000000000005 mm
Heaviest single-day rainfall: 26.2 mm


In [63]:
rainy_avg = temps[rain > 0].mean()
dry_avg = temps[rain == 0].mean()

print(f"Average temperature on rainy days: {rainy_avg:.2f} °C")
print(f"Average temperature on dry days: {dry_avg:.2f} °C")

Average temperature on rainy days: 24.14 °C
Average temperature on dry days: 23.95 °C


In [64]:
hot_and_rainy = (temps >= 26) & (rain > 0)

print("Hot and rainy days:", hot_and_rainy.sum())

Hot and rainy days: 5


In [65]:
high_humidity = hum >= 70

print("Days with humidity >= 70%:", high_humidity.sum())

Days with humidity >= 70%: 16


In [66]:
#PART 5 Anomalies & Trends

In [67]:
anomaly = temps - temps.mean()

In [68]:
above_mean = anomaly > 3

print("Dates more than 3°C above the mean:")
print(dates[above_mean])

Dates more than 3°C above the mean:
[46272 46274 46276 46277 46279 46281 46282 46284 46287]


In [69]:
below_mean = anomaly < -3

print("Days more than 3°C below the mean:", below_mean.sum())

Days more than 3°C below the mean: 8


In [70]:
first_30_avg = temps[:30].mean()
last_30_avg = temps[30:].mean()

print(f"First 30 days average: {first_30_avg:.2f} °C")
print(f"Last 30 days average: {last_30_avg:.2f} °C")

# Kigali got cooler because the average temperature decreased.

First 30 days average: 26.21 °C
Last 30 days average: 21.78 °C


In [71]:
# PART OF Bonus

In [72]:
wettest_days = sorted(
    zip(dates, rain),
    key=lambda x: x[1],
    reverse=True
)

print("Top 5 wettest days:")

print(*wettest_days[:5], sep="\n")

Top 5 wettest days:
(46272, 26.2)
(46279, 25.1)
(46286, 24.0)
(46270, 22.9)
(46293, 22.9)


In [74]:
#PART 6

NumPy makes calculations easier by doing the repeated steps for us. For example, temps.mean() gives the average without needing a loop AND ALSO  Learning loops first helped me understand what NumPy is doing behind the scenes OF CODES Thanks

AND Also   One important insight from my analysis is that Kigali became considerably cooler 
during the second half of the 60-day period. The average temperature decreased 
from about 26.21°C during the first 30 days to about 21.78°C during the last 30 
days. This shows a noticeable cooling trend in the weather data.
